File: challenges-in-python/merge_intervals/merge_intervals.py
# Solving the "Merge Intervals" Challenge
# Challenge Description:
# Given an array of intervals where intervals[i] = [starti, endi], 
# merge all overlapping intervals, 
# and return an array of the non-overlapping intervals 
# that cover all the intervals in the input.

# Approach:
# Sort the intervals based on their start times.
# Iterate through the intervals, and if an interval 
# overlaps with the previous one (its start time is less than or equal 
# to the end time of the previous interval), merge them. Otherwise, add it to the output.

def merge(intervals: list[list[int]]) -> list[list[int]]:
    # if intervals is empty return empty list
    if not intervals:
        return []
    
    # Sort the intervals based on their start times
    intervals.sort(key=lambda x: x[0])

    # Initialize merged list with the first interval
    merged = [intervals[0]]
    for i in range(1, len(intervals)):
        # if the current interval overlaps with the last merged interval
        # merge them by updating the end time of the previous interval
        if intervals[i][0] <= merged[-1][1]:
            # merge them by updating the end time of the previous interval
            merged[-1][1] = max(intervals[i][1], merged[-1][1])
        else:
            # add it to the output
            merged.append(intervals[i])
    
    return merged

# Testing the function
def test_merge():
    assert merge([[1,3],[2,6],[8,10],[15,18]]) == [[1,6],[8,10],[15,18]]
    assert merge([[1,4],[4,5]]) == [[1,5]]
    assert merge([[1,4],[5,6]]) == [[1,4], [5,6]]
    # empty list
    assert merge([]) == []
    # single interval
    assert merge([[1,4]]) == [[1,4]]

test_merge()

File: challenges-in-rust/src/merge_intervals/merge_intervals.rs
// Solving the "Merge Intervals" Challenge
// Problem Statement
// Given an array of intervals where intervals[i] = [start_i, end_i], 
// merge all overlapping intervals and return an array of the non-overlapping intervals 
// that cover all the intervals in the input.

// Input: intervals = [[1,3],[2,6],[8,10],[15,18]]
// Output: [[1,6],[8,10],[15,18]]

// Implementation:
// Time Complexity: O(n log n) 
// Space Complexity: O(n) 
pub fn merge(mut intervals: Vec<Vec<i32>>) -> Vec<Vec<i32>> {
    // Check if the intervals are empty
    if intervals.is_empty() {
        return intervals;
    }
    // Sort the intervals by the start time
    intervals.sort_by(|a, b| a[0].cmp(&b[0]));
    let mut result: Vec<Vec<i32>> = Vec::new();
    let mut current_interval = intervals[0].clone(); // Clone the first interval
    for interval in intervals {
        if interval[0] <= current_interval[1] { // Check if the interval overlaps
            current_interval[1] = current_interval[1].max(interval[1]); // Merge the interval
        } else {
            result.push(current_interval.clone());
            current_interval = interval;
        }
    }
    result.push(current_interval);
    result
}

// Test Cases:
#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_merge_no_overlap() {
        let intervals = vec![vec![1, 3], vec![4, 6], vec![8, 10]];
        let expected = vec![vec![1, 3], vec![4, 6], vec![8, 10]];
        assert_eq!(merge(intervals), expected);
    }

    #[test]
    fn test_merge_with_overlap() {
        let intervals = vec![vec![1, 3], vec![2, 6], vec![8, 10], vec![15, 18]];
        let expected = vec![vec![1, 6], vec![8, 10], vec![15, 18]];
        assert_eq!(merge(intervals), expected);
    }

    #[test]
    fn test_merge_subinterval() {
        let intervals = vec![vec![1, 4], vec![2, 3]];
        let expected = vec![vec![1, 4]];
        assert_eq!(merge(intervals), expected);
    }

    #[test]
    fn test_merge_single_interval() {
        let intervals = vec![vec![5, 7]];
        let expected = vec![vec![5, 7]];
        assert_eq!(merge(intervals), expected);
    }

    #[test]
    fn test_merge_empty() {
        let intervals: Vec<Vec<i32>> = Vec::new();
        let expected: Vec<Vec<i32>> = Vec::new();
        assert_eq!(merge(intervals), expected);
    }
}